src/cargo/core/workspace.rs

```diff
@@ -506,13 +528,43 @@ impl<'gctx> Workspace<'gctx> {
     /// Returns the root `[patch]` section of this workspace.
     ///
     /// This may be from a virtual crate or an actual crate.
-    pub fn root_patch(&self) -> CargoResult<HashMap<Url, Vec<Dependency>>> {
-        let from_manifest = match self.root_maybe() {
+    pub fn root_patch(&self) -> CargoResult<HashMap<Url, Vec<(Dependency, PathOrDefinition)>>> {
```

Comment:
I'm a bit mixed on tuple vs struct.

Reply:
We could have a struct like:

```
#[derive(Clone, Debug)]
pub struct Patch {
    dep: Dependency,
    location: PathOrDefinition,
}
```
and have the `PathOrDefinition::get_location_string` implemented for it instead. But I'm not sure what would be the other benefits, as we would still need to destructure it in a lot of places or keep using `iter().map(|patch| patch.dep)` instead. But I'm open for better ideas.